lxd/instance/drivers/driver_qemu.go

```diff
@@ -7427,21 +7442,65 @@ func (d *qemu) DeviceEventHandler(runConf *deviceConfig.RunConfig) error {
 }
 
 // vsockID returns the vsock Context ID for the VM.
-func (d *qemu) vsockID() int {
-	// We use the system's own VsockID as the base.
-	//
-	// This is either "2" for a physical system or the VM's own id if
-	// running inside of a VM.
-	//
-	// To this we add 1 for backward compatibility with prior logic
-	// which would start at id 3 rather than id 2. Removing that offset
-	// would cause conflicts between existing VMs until they're all rebooted.
-	//
-	// We then add the VM's own instance id (1 or higher) to give us a
-	// unique, non-clashing context ID for our guest.
+func (d *qemu) vsockID() (int, error) {
+	// Return an already existing Context ID.
+	existingVsockID, ok := d.localConfig["volatile.vsock_id"]
+	if ok {
+		vsockID, err := strconv.Atoi(existingVsockID)
+		if err != nil {
+			return 0, fmt.Errorf("Failed to convert Context ID: %w", err)
+		}
 
-	info := DriverStatuses()[instancetype.VM].Info
-	return info.Features["vhost_vsock"].(int) + 1 + d.id
+		return vsockID, nil
+	}
+
+	instanceUUID := uuid.Parse(d.localConfig["volatile.uuid"])
+	if instanceUUID == nil {
+		return 0, fmt.Errorf("Instance does not have a UUID")
+	}
+
+	// Pick the first 4 bytes of the random UUID (v4) to create an uint32 representation.
+	// Since the instances UUID already exist, we don't need to create another random value.
+	firstBytes := binary.LittleEndian.Uint32(instanceUUID[:4])
```

Comment:
@roosterfish This seems like a potential candidate for `util.GetStableRandomGenerator()`.

E.g.

```go
r, err := util.GetStableRandomGenerator(instanceUUID)
if err != nil {
    return fmt.Errorf("Failed generating stable random seed from instance UUID %q: %w", instanceUUID, err)
}
```

Then in a loop up until max number of attempts:

```go
candidateVsockID := uint32(r.Int31())
```

This will then use the full uuid as the basis for a stable random number generator, so it will optimistically use the same ID if possible, but if not possible, then it will allow you to conveniently get additional random numbers to try.

Reply:
Actually `*rand.Rand` already comes with a `r.UInt32()` :) It definitely looks cleaner in the code. But I am wondering which limit for the loop we can now apply? Now you could just iterate from `x` to `x-1` ranging over all the possible values for `uint32` by causing an overflow.

Leaving the loop open without applying a limit would be my preference if this is acceptable.

Reply:
I think we should bound it by time rather than number of iterations. Shall we say 5s?

Reply:
Yes, good idea. Fixed.